src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraph.java

```diff
@@ -0,0 +1,464 @@
+        for (final String child : cgs.getAllDescendentsMap().get(key)) {
+            CacheNode oldChild = nodesMap.get(child);
+            /* all of the desendents are computable nodes and can be safely upcasted */
+            updatedNodesMap.put(child, ((ComputableCacheNode)oldChild).duplicateWithUpdatedCacheStatus(false));
+        }
+        return updatedNodesMap;
+    }
+
+    /**
+     * Returns a reference to the value of a given node
+     *
+     * Note: this function is purposefully meant to be light:
+     *
+     * (1) it does not update out-of-date cacheable computable nodes, and
+     * (2) it does not evaluate non-cacheable computable nodes.
+     *
+     * @param nodeKey the key of the node
+     * @return value of the node
+     * @throws IllegalStateException if a cached node is out of date, or if the value of a primitive node is
+     *                               not initialized
+     * @throws IllegalArgumentException if the node does not exist
+     */
+    public Duplicable getValueDirect(@Nonnull final String nodeKey)
+            throws IllegalStateException, IllegalArgumentException {
+        assertNodeExists(Utils.nonNull(nodeKey, "The key of the node can not be null."));
+        return nodesMap.get(nodeKey).getValue(EMPTY_MAP);
+    }
+
+    /**
+     *
+     * @param nodeKey key of the node
+     * @return value of the node
+     * @throws IllegalArgumentException if the node does not exist
+     */
+    public Duplicable getValueWithRequiredEvaluations(@Nonnull final String nodeKey)
+            throws IllegalStateException, IllegalArgumentException {
+        assertNodeExists(Utils.nonNull(nodeKey, "The key of the node can not be null."));
+        return accumulateEvaluationsForDepthSortedNodeList(cgs.getEvalStrategyByEvaluatedNode(nodeKey), true).get(nodeKey);
+    }
+
+    /**
+     * This method calculates and accumulates the values of a list of depth-sorted nodes (from low depth to high depth).
+     * Compute-on-demand nodes are always evaluated. Caching nodes are only evaluated if {@code evalOnOutOfDateCachingNodes}
+     * is true.
+     *
+     * @param depthSortedNodes depth-sorted list of nodes
+     * @param evalOutdatedCachingNodes evaluate
+     * @return
+     */
+    private Map<String, Duplicable> accumulateEvaluationsForDepthSortedNodeList(@Nonnull final List<String> depthSortedNodes,
+                                                                                final boolean evalOutdatedCachingNodes) {
+        final Map<String, Duplicable> accumulatedValues = new HashMap<>();
+        for (final String nodeKey : depthSortedNodes) {
+            if (nodesMap.get(nodeKey).isPrimitive()) {
+                Duplicable value;
+                try {
+                    value = nodesMap.get(nodeKey).getValue(accumulatedValues);
+                } catch (IllegalStateException e) {
+                    value = null;
```

Comment:
This part of the code is _a lot_ simpler now:

```
    /**
     * This method computes and accumulates the values on a list of topologically ordered set of nodes (here meaning
     * from low depth to high depth).
     *
     * Along the way and en route to the output map,
     *
     *   - The value of a primitive node is passed by reference
     *   - A non-caching computable node is always computed
     *   - A caching computable node is only computed if it has no cached value or its stored valued is not current
     *
     * Note: this method does not check whether {@code topologicallyOrderedNodeKeys} is actually topologically ordered.
     *
     * @param topologicallyOrderedNodeKeys depth-sorted list of nodes
     * @throws ComputableNodeFunction.ParentValueNotFoundException if a parent value required for a computation function
     *         is not found; it can be thrown if {@code topologicallyOrderedNodeKeys} is not truly topologically ordered
     * @throws ComputableCacheNode.ExternallyComputableNodeValueUnavailableException if the value of an externally computable
     *         node is out of date or not initialized
     * @throws PrimitiveCacheNode.PrimitiveValueNotInitializedException if the value of a required primitive node is not
     *         initialized
     * @return a map from node keys to their values accumulated during computation
     */
    private Map<String, Duplicable> evaluateInTopologicalOrder(@Nonnull final List<String> topologicallyOrderedNodeKeys) {
        final Map<String, Duplicable> accumulatedValues = new HashMap<>();
        for (final String nodeKey : topologicallyOrderedNodeKeys) {
            accumulatedValues.put(nodeKey, nodesMap.get(nodeKey).get(accumulatedValues));
        }
        return accumulatedValues;
    }
```